File: scala/leetcode/src/main/scala/leetcode/editor/cn/[206]反转链表.scala
package leetcode.editor.cn
object ReverseLinkedList {
//leetcode submit region begin(Prohibit modification and deletion)
  /** Definition for singly-linked list. class ListNode(_x: Int = 0, _next: ListNode = null) { var next: ListNode = _next var x: Int = _x }
    */
  object Solution {
    def reverseList(head: ListNode): ListNode =
      if head == null
      then head
      else reverseN(head, Int.MaxValue)

    @annotation.tailrec
    private def reverseN(head: ListNode, n: Int, acc: ListNode = null): ListNode =
      if n == 0 || head == null
      then acc
      else
        val next = head.next
        head.next = acc
        reverseN(next, n - 1, head)
  }
//leetcode submit region end(Prohibit modification and deletion)

}

File: cpp/leetcodecn/[206]反转链表.cpp
#include <vector>
#include "util.h"
#include <map>
#include <string>

using namespace std;

//leetcode submit region begin(Prohibit modification and deletion)
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode *reverseList(ListNode *head, ListNode *acc = nullptr) {
        if (head == nullptr) return acc;
        else {
            auto next = head->next;
            head->next = acc;
            return reverseList(next, head);
        }
    }
};
//leetcode submit region end(Prohibit modification and deletion)


int main(int argv, char *argc[]) {
}